source/structuring.rst

```diff
@@ -14,402 +14,545 @@ functions is better than copying-and-pasting duplicate bits of code
+.. schema_example::
 
-``$ref`` can resolve to a URI that references another file, so if you prefer to
-include your definitions in separate files, you can also do that.  For
-example::
+    {
+      "$id": "/schemas/address",
 
-    { "$ref": "definitions.json#/address" }
+      "type": "object",
+      "properties": {
+        "street_address": { "type": "string" },
+        "city": { "type": "string" },
+        "state": { "type": "string" }
+      },
+      "required": ["street_address", "city", "state"]
+    }
+
+No matter which URI is used to retrieve this schema, the canonical URI
+will be ``https://example.com/schemas/address``, which is the result
+of the ``$id`` URI-reference resolving against the `retrieval-uri`.
+Implementations will use this canonical URI instead of the
+`retrieval-uri` whenever a URI-reference needs to be resolved.
 
-would load the address schema from another file residing alongside
-this one.
+However, using a relative reference when setting a canonical URI can
+be problematic. For example, we couldn't use this schema as a schema
+literal because there would be no `retrieval-uri` and you can't
+resolve a relative reference against nothing. For this and other
+reasons, it's recommended that you always use an absolute URI when
+declaring a canonical URI.
 
-Now let's put this together and use our address schema to create a
-schema for a customer:
+The identifier of the following schema will always be
+``https://example.com/schemas/address`` no matter what the
+`retrieval-uri` was or if it's used as an anonymous schema.
 
 .. schema_example::
 
     {
-      "$schema": "http://json-schema.org/draft-07/schema#",
-
-      "definitions": {
-        "address": {
-          "type": "object",
-          "properties": {
-            "street_address": { "type": "string" },
-            "city":           { "type": "string" },
-            "state":          { "type": "string" }
-          },
-          "required": ["street_address", "city", "state"]
-        }
-      },
+      "$id": "https://example.com/schemas/address",
 
       "type": "object",
-
       "properties": {
-        "billing_address": { "$ref": "#/definitions/address" },
-        "shipping_address": { "$ref": "#/definitions/address" }
-      }
-    }
-    --
-    {
-      "shipping_address": {
-        "street_address": "1600 Pennsylvania Avenue NW",
-        "city": "Washington",
-        "state": "DC"
+        "street_address": { "type": "string" },
+        "city": { "type": "string" },
+        "state": { "type": "string" }
       },
-      "billing_address": {
-        "street_address": "1st Street SE",
-        "city": "Washington",
-        "state": "DC"
-      }
+      "required": ["street_address", "city", "state"]
     }
 
 .. note::
+   The behavior when setting a canonical URI that contains a URI
+   fragment is undefined and should not be used because
+   implementations may treat them differently.
```

Comment:
Remember that at this point UJS is still documenting draft-07 where fragments were not forbidden yet. This will be updated later with the Draft 2019-09 changes this will be moved to a "draft_specific" box.

Reply:
Yes. Sorry. Totally right.